### ISSUE TITLE
Document notification coalescing

### ISSUE THREAD

**Issue by icanzilb** [realm/realm-cocoa#3435]
## Goals

This is what the API docs say (thanks for fixing this) about `Results.addNotificationBlock(_)`:

> The block will be asynchronously called with the initial results, and then called again after each write transaction which causes the results to change.

I have code that creates a `Results` object and then writes one object that changes the results. In this case I'd expect my notification block to be called two times. The block however is called just one time.

## Expected Results

The block to be called for the initial result set

## Actual Results

The notification block is not called for the initial result set.

## Steps to Reproduce

```swift
Here's the complete source code to reproduce:

import UIKit
import RealmSwift

class Message: Object {
    dynamic var id = 0
    dynamic var text = ""
}

class ViewController: UIViewController {

    var token: NotificationToken?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let realm = try! Realm()
        try! realm.write {
            realm.deleteAll()
        }
        
        let hello1 = Message()
        hello1.text = "Hello!"

        try! realm.write {
            realm.add(hello1)
        }
        
        let feed = realm.objects(Message)
        
        print("initial count: \(feed.count)")
        token = feed.addNotificationBlock {newResults, _ in
            print("notification count: \(newResults!.count)")
        }

        let hello = Message()
        hello.text = "Hello!"
        
        try! realm.write {
            realm.add(hello)
        }
        print("changed count: \(feed.count)")
    }
    
}
```

Here's the output from the print statements:

![image](https://cloud.githubusercontent.com/assets/633862/14520491/1a47e748-0225-11e6-9ecf-b59e58c34d63.png)

## Version of Realm and Tooling

Realm version: 0.98.7

Xcode version: 7,3

iOS/OSX version: latest

Dependency manager + version: Cocoapods 0.39

**Comment by jpsim** [comment]
This is actually behaving as expected, but we're missing some documentation on the matter.

Sometimes, write transaction notifications get *coalesced*, for example when multiple write transactions occur within a single iteration of the runloop on which the notification block was added.

The documentation does state that the notification block is initially called *asynchronously*, but we fail to mention this coalescing behavior. I'll update this ticket's title to reflect that the "fix" here is actually to improve our documentation.

**Comment by jpsim** [comment]
For example, if you triggered the `realm.add(_:)` write in a `dispatch_after` block (within a reasonable delay), the block would actually be invoked twice.

**Comment by icanzilb** [comment]
I was already thinking that might be the case but wanted to make sure whether it's the expected behavior or not. It's worth a mention in the docs 👍

**Comment by mrackwitz** [comment]
This is going to be fixed by #3359.
